custom_components/ha_washdata/profile_store.py

```
@@ -5623,17 +6047,354 @@ async def async_import_data(self, payload: dict[str, Any]) -> dict[str, Any]:
+        ref_id_pool: set[Any] = {c.get("id") for c in self._data.get("reference_cycles", []) if isinstance(c, dict)}
+        past_id_pool: set[Any] = {c.get("id") for c in self._data.get("past_cycles", []) if isinstance(c, dict)}
+        # Reference-cycle dedup: skip re-importing a reference cycle already held so a
+        # repeated import stays idempotent (otherwise envelopes get double-weighted and
+        # cycle_count inflates). Keyed on the stable source id stamped into meta.source.
+        existing_ref_sources: set[str] = {
+            str((c.get("meta") or {}).get("source"))
+            for c in self._data.get("reference_cycles", [])
+            if isinstance(c, dict) and (c.get("meta") or {}).get("source")
+        }
+
+        def _bare_store_id(raw_id: Any) -> str:
```

Ruff (0.15.21)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `raw_id`

(ANN401)